src/Rel8/Table/Verify.hs

```diff
@@ -0,0 +1,642 @@
+  in CheckEnv tblMap castMap
+ where
+  fetchSchema :: HS.Statement () ([PGTable Result], [Cast Result])
+  fetchSchema = run1 $ select $ liftA2 (,) fetchTables fetchCasts
+
+
+-- |@'SomeTableSchema'@ is used to allow the collection of a variety of different
+-- @TableSchema@s under a single type, like:
+--
+-- @
+-- userTable :: TableSchema (User Name)
+-- orderTable :: TableSchema (Order Name)
+--
+-- tables :: [SomeTableSchema]
+-- tables = [SomeTableSchema userTable, SomeTable orderTable]
+-- @
+--
+-- This is used by @'schemaErrors'@ to more conveniently group every table an
+-- application relies on together, for typechecking the postgresql schemas
+-- together in a single batch.
+data SomeTableSchema where
+    -- The ToExpr constraint isn't used here, but can be used to read from the
+    -- SomeTableSchema, which can be useful to combine the type checking with more
+    -- thorough value-level checking of the validity of existing rows in the
+    -- table.
+    SomeTableSchema
+        :: (ToExprs (k Expr) (GFromExprs k), Rel8able k)
+        => TableSchema (k Name) -> SomeTableSchema
+
+-- |@'getSchemaErrors'@ checks whether the provided schemas have correct postgresql
+-- column names and types to allow reading and writing from their equivalent Haskell
+-- types, returning a list of errors if that is not the case. The function does not
+-- crash on an encountered bug, instead leaving it to the caller to decide how
+-- to respond. A schema is valid if:
+--
+-- 1. for every existing field, the types match
+-- 2. all non-nullable columns are present in the hs type
+-- 3. no nonexistent columns are present in the hs type
+-- 4. no two columns in the same schema share the same name
+--
+-- It's still possible for a valid schema to allow invalid data, for instance,
+-- if using an ADT, which can introduce restrictions on which values are allowed
+-- for the column representing the tag, and introduce restrictions on which
+-- columns are non-null depending on the value of the tag. However, if the
+-- schema is valid rel8 shouldn't be able to write invalid data to the table.
+--
+-- However, it is possible for migrations to cause valid data to become invalid,
+-- in ways not detectable by this function, if the migration code changes the
+-- schema correctly but doesn't handle the value-level constraints correctly, so
+-- it is a good idea to both read from the tables and check the schema for errors
+-- in a transaction during the migration. This former will catch value-level
+-- bugs, while the latter will help ensure the schema is set up correctly to
+-- be able to insert new data.
+--
+-- This function does nothing to check that the conflict target of an @Upsert@
+-- are valid for the schema, nor can it prevent invalid uses of @unsafeDefault@.
+-- However, it should be enough to catch the most likely errors.
+getSchemaErrors :: [SomeTableSchema] -> HS.Statement () (Maybe Text)
```

Comment:
Do we want Text or String here. Some of the other functions return String but this returns Text. I don't think it makes much of a difference but I wanted to check whether you wanted it to be consistent.